### PR TITLE
on_exception (#347)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
from src.taipy.gui import Gui, State, notify

md = """
#on_exception test
<|Generate an exception|button|on_action=gen_exc|>
"""

def gen_exc(state: State):
    state.toto = 10

def on_exception(state: State, function_name: str, exc: Exception):
    notify(state, "E", f"Excetion occured in {function_name}: {exc}")

Gui(md).run(use_reloader=False)

### CODE REVIEW

File: src/taipy/gui/gui.py [hunk 1]
```diff
@@ -96,6 +96,9 @@ class Gui:
         on_init (Callable): The function that is called on the first connection of a new user.<br/>
             It defaults to the `on_init()` global function defined in the Python
             application.
+        on_exception (Callable): The function that is called an exception occurs on user code.<br/>
+            It defaults to the `on_exception()` global function defined in the Python
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py:105]: What if there's no such `on_exception()`?

File: src/taipy/gui/gui.py [hunk 2]
```diff
@@ -1186,9 +1191,19 @@ def __init_route(self):
                 try:
                     self._call_function_with_state(self.on_init, [])
                 except Exception as e:
-                    warnings.warn(f"Exception on on_init execution \n{e}")
+                    if not self.__call_on_exception("on_init", e):
+                        warnings.warn(f"Exception on on_init execution \n{e}")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: -> "Exception raised in 'on_init'\n{e}"

File: src/taipy/gui/gui.py [hunk 3]
```diff
@@ -1186,9 +1191,19 @@ def __init_route(self):
...
         return self._render_route()
 
+    def __call_on_exception(self, function_name: str, exception: Exception) -> bool:
+        if hasattr(self, "on_exception") and callable(self.on_exception):
+            try:
+                self.on_exception(self.__get_state(), str(function_name), exception)
+            except Exception as e:
+                warnings.warn(f"Exception occured on on_exception execution\n{e}")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: -> "Exception raised in 'on_exception'\n{e}"

**Review by FabienLelaquais — APPROVED** [review]
Minor changes.
Well done!

**Review by dinhlongviolin1 — APPROVED** [review]

**Pull request merged**